crates/ra_ide/src/extend_selection.rs

```diff
@@ -88,6 +105,67 @@ fn try_extend_selection(root: &SyntaxNode, range: TextRange) -> Option<TextRange
     node.parent().map(|it| it.text_range())
 }
 
+fn extend_tokens_from_range(
+    db: &RootDatabase,
+    root: &InFile<&SyntaxNode>,
+    macro_call: ast::MacroCall,
+    original_range: TextRange,
+) -> Option<TextRange> {
+    let analyzer = hir::SourceAnalyzer::new(db, root.clone(), None);
+    let expansion = analyzer.expand(db, root.with_value(&macro_call))?;
+
+    // compute original mapped token range
+    let range = macro_call
+        .syntax()
+        .descendants_with_tokens()
+        .filter_map(|n| match n {
+            NodeOrToken::Token(token) if token.text_range().is_subrange(&original_range) => {
+                expansion
+                    .map_token_down(db, root.with_value(&token))
+                    .map(|node| node.value.text_range())
+            }
+            _ => None,
+        })
+        .fold1(|x, y| union_range(x, y))?;
+
+    let src = db.parse_or_expand(expansion.file_id())?;
+    let parent = shallow_node(&find_covering_element(&src, range))?.parent()?;
+
+    // compute parent mapped token range
+    let range = macro_call
+        .syntax()
+        .descendants_with_tokens()
+        .filter_map(|n| match n {
+            NodeOrToken::Token(token) => {
+                expansion.map_token_down(db, root.with_value(&token)).and_then(|node| {
+                    if node.value.text_range().is_subrange(&parent.text_range()) {
+                        Some(token.text_range())
+                    } else {
+                        None
+                    }
+                })
+            }
+            _ => None,
+        })
+        .fold1(|x, y| union_range(x, y))?;
+
+    if original_range.is_subrange(&range) && original_range != range {
+        Some(range)
+    } else {
+        None
+    }
+}
+
+fn union_range(range: TextRange, r: TextRange) -> TextRange {
```

Comment:
Would it makes sense to move this to a `TextRange::union(a, b)`? I could imagine it being useful for other things

Reply:
Indeed, but `TextRange` is in https://github.com/rust-analyzer/text_unit, I would make a PR in future .